mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -375,6 +493,177 @@ struct BackpropOpPattern : public ConvertOpToLLVMPattern<BackpropOp> {
+        for (Type argType : argTypes) {
+            llvmArgTypes.append(2, argType);
+        }
+    }
+
+    /// Generate an augmented forward pass, which computes the original value of the `qnode` in
+    /// addition to any data that the gradient of this qnode needs. Currently, the gradient does not
+    /// require any intermediate data, so the augmented forward pass just calls + returns the result
+    /// of the regular QNode.
+    ///
+    /// Enzyme additionally expects this function to take shadow pointers of forward pass arguments,
+    /// which are currently unused. For more detail, see:
+    /// https://enzyme.mit.edu/getting_started/CallingConvention#custom-gradients
+    func::FuncOp genAugmentedForward(func::FuncOp qnode, OpBuilder &builder) const
+    {
+        std::string augmentedName = (qnode.getName() + ".augfwd").str();
+        auto augmentedForward = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+            qnode, builder.getStringAttr(augmentedName));
+        if (augmentedForward) {
+            return augmentedForward;
+        }
+        assert(qnode.getNumResults() == 0 && "Expected QNode to be in destination-passing style");
+        MLIRContext *ctx = builder.getContext();
+        OpBuilder::InsertionGuard insertionGuard(builder);
+        builder.setInsertionPointAfter(qnode);
+        // The tape type is a null pointer because we don't need to pass any data from the forward
+        // pass to the reverse pass.
+        auto tapeType = LLVM::LLVMPointerType::get(ctx);
+        SmallVector<Type> argTypes;
+        convertCustomGradArgumentTypes(qnode.getArgumentTypes(), argTypes);
+        augmentedForward = builder.create<func::FuncOp>(
+            qnode.getLoc(), augmentedName, FunctionType::get(ctx, argTypes, {tapeType}));
+        augmentedForward.setPrivate();
+        Location loc = qnode.getLoc();
+
+        Block *entry = augmentedForward.addEntryBlock();
+        builder.setInsertionPointToStart(entry);
+
+        // Every other argument is a shadow
+        SmallVector<Value> arguments;
+        for (unsigned i = 0; i < qnode.getNumArguments(); i++) {
+            arguments.push_back(augmentedForward.getArgument(i * 2));
+        }
+
+        builder.create<func::CallOp>(loc, qnode, arguments);
+        Value tape = builder.create<LLVM::NullOp>(loc, tapeType);
+        builder.create<func::ReturnOp>(loc, tape);
+        return augmentedForward;
+    }
+
+    /// Given a `qnode` in destination passing style and a function that computes its quantum
+    /// derivative with respect to gate parameters `qgradFn`, generate an Enzyme-compatible custom
+    /// gradient function that computes the derivative of the QNode in the context of a larger,
+    /// Enzyme-driven hybrid gradient computation.
+    ///
+    /// To do this, it takes the incoming cotangent (expressed as the shadow of the QNode result)
+    /// and applies the chain rule via multiplication with the quantum derivative to produce the
```

Comment:
Is it really the chain rule and not just the multiplication?: $x^t \times \nabla J$

Reply:
The chain rule explains *why* we multiply partial derivatives together in AutoDiff. It's the reason that backwards propagation of vector-Jacobian products (and forward propagation of Jacobian-vector products) produce the correct total derivative.